### PR TITLE
Remove table and clean up redundant state info when scanning newly added table

### PR DESCRIPTION
**Pull request by qidian99**
As mentioned in [issue@913](https://github.com/ververica/flink-cdc-connectors/issues/913), when a task is stopped and its configuration changed, there might be redundant table information remaining in the states of SplitAssigner and SourceReader. 

The current functionality of scanning newly added tables will removing and cleaning up table info in the previous savepoint according to the new configuration.

The changes proposed in this PR is as follows:

1. In MySqlSplitAssigner, it will apply the new filter in the new configuration and remove redundant table info
2. In MySqlSourceReader, it will sieve out splits that do not match the new filter.

### CODE REVIEW

**Review by zhjwpku — APPROVED** [review]

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java [hunk 1]
```diff
@@ -169,14 +171,44 @@ private void captureNewlyAddedTables() {
...
                     remainingTables.addAll(newlyAddedTables);
-                    if (isAssigningFinished(assignerStatus)) {
+
+                    // remove unmatched tables from already processed tables since the abnormal
+                    // tables might be added back in the future
+                    alreadyProcessedTables.removeIf(tableId -> !tableFilter.isIncluded(tableId));
+
+                    // only suspend the assigner if there's new tables added and assigner has
+                    // finished
+                    if (isAssigningFinished(assignerStatus) && !newlyAddedTables.isEmpty()) {
```
Comment (zhjwpku) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java]: `!newlyAddedTables.isEmpty()` is called twice, can we split the removed and added logic to make it more clear?

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java [hunk 2]
```diff
@@ -169,13 +170,42 @@ private void captureNewlyAddedTables() {
...
+                // case 2: there are new tables to add
                 if (!newlyAddedTables.isEmpty()) {
                     // if job is still in snapshot reading phase, directly add all newly added
                     // tables
                     LOG.info("Found newly added tables, start capture newly added tables process");
+
+                    // add new tables
                     remainingTables.addAll(newlyAddedTables);
+
+                    // only suspend the assigner if there's new tables added and assigner has
```
Comment (zhjwpku) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:207]: only suspend the assigner if the assigner in finished status

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java (hunk 1 above)
Comment (qidian99) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java]: Changed the code to remove tables first and then add new tables.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java (hunk 2 above)
Comment (qidian99) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java]: Comment updated.

**Comment by zhjwpku** [comment]
LGTM, @leonardBang pls take a look at this PR.

**Pull request closed without merging**